aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -119,279 +121,390 @@ pub struct AggregatorSnapshot {
+            }
         }
-        self.record_success();
         Ok(())
     }
 
-    /// Implements logic for reading the value of an aggregator. As a
-    /// result, the aggregator knows it value (i.e. its state changes to
-    /// `Data`).
-    pub fn read_and_materialize(
+
+    /// Implements logic for doing a "cheap read" of an aggregator.
+    /// This means that we query the MVHashmap for the last committed value
+    /// of the aggregator, and store it in the `speculative_start_value`, with
+    /// `speculative_source` as `LastCommitted`.
+    pub fn read_last_committed_aggregator_value(
         &mut self,
         resolver: &dyn AggregatorResolver,
-        id: &AggregatorID,
     ) -> PartialVMResult<u128> {
-        // If aggregator has already been read, return immediately.
-        if self.state == AggregatorState::Data {
-            return Ok(self.value);
-        }
-
-        // Otherwise, we have a delta and have to go to storage and apply it.
-        // In theory, any delta will be applied to existing value. However,
-        // something may go wrong, so we guard by throwing an error in
-        // extension.
-        let value_from_storage = resolver.resolve_aggregator_value(id).map_err(|e| {
-            extension_error(format!("Could not find the value of the aggregator: {}", e))
-        })?;
-
-        // Validate history and apply the delta.
-        self.validate_history(value_from_storage)?;
         match self.state {
-            AggregatorState::PositiveDelta => {
-                self.value = addition(value_from_storage, self.value, self.max_value)?;
+            AggregatorState::Data { value } => {
+                // If aggregator knows the value, return it.
+                Ok(value)
             },
-            AggregatorState::NegativeDelta => {
-                self.value = subtraction(value_from_storage, self.value)?;
-            },
-            AggregatorState::Data => {
-                unreachable!("Materialization only happens in Delta state")
+            AggregatorState::Delta { speculative_start_value, speculative_source, delta, history } => {
+                // If we performed a "cheap read" or "expensive read" operation before, use it.
+                if speculative_source != SpeculativeValueSource::Default {
+                    return match delta {
+                        DeltaValue::Positive(value) => Ok(addition(speculative_start_value, value, self.max_value)?),
+                        DeltaValue::Negative(value) => Ok(subtraction(speculative_start_value, value)?),
+                    };
+                }
+                // Otherwise, we have to go to storage and read the value.
+                let value_from_storage = resolver.resolve_last_committed_aggregator_value(&self.id).map_err(|e| {
+                    extension_error(format!("Could not find the value of the aggregator: {}", e))
+                })?;
+        
+                // Validate history and apply the delta.
```

Comment:
if we are only now doing cheap read, history should be empty, maybe we should check that instead?